### DISCUSSION TITLE
nested struct retrieval

### DISCUSSION THREAD

**Discussion by pnathan** [pulumi/pulumi#9138] [Pulumi in Practice]
Hi,

Trying to access a nested element:

        import "github.com/pulumi/pulumi-aws/sdk/v4/go/aws/iam"

	tp := ekscluster.Core.OidcProvider().ToOpenIdConnectProviderOutput().ApplyT(func(x *iam.OpenIdConnectProvider) pulumi.StringArrayOutput {
		return x.ThumbprintLists
	})

this fails, because apparently this is cast to AnyOutput. 

What I really would like is to be able to do something like ekscluster.Core.OidcProvider.ThumbprintLists.GetOutput  // gets a stringarray.

not sure what to do.

**Comment by rawkode** [comment]
What are you trying to do next with `tp`? Perhaps we can recommend a next step, but we don't have enough information to help just yet.

**Reply by pnathan** [reply]
@rawkode - sure! Sorry about that.

```
oidc, err := aws.NewOpenIdConnectProvider(ctx, "Test1", &aws.OpenIdConnectProviderArgs{
			ClientIdLists: pulumi.StringArray{
				pulumi.String("sts.amazonaws.com"),
			},
			ThumbprintLists: tp,
			Url:             url,
		})
```

**Reply by pnathan** [reply]
The essence of what I need to do do is something analogous to this rather horrible Haskell code.

```
module Main where

import Control.Applicative


data Identity = Identity {
  url :: String,
  tp  :: String
  } deriving Show

data Cluster = Cluster {
  name :: String,
  identity :: Identity
  } deriving Show

data Oidc = Oidc {
  address :: String,
  thumbprint :: String,
  id :: Int
  } deriving Show

data Pulumi a = Pulumi a deriving Show

instance Functor Pulumi where
  fmap f (Pulumi x) = Pulumi (f x)

instance Applicative Pulumi where
  pure = Pulumi
  (Pulumi f) <*> x = f <$> x

instance Monad Pulumi where
      return x = Pulumi x
      Pulumi x >>= f = f x

main :: IO ()
main = do
  let c = Cluster { name = "C",
                    identity = Identity
                               {
                                 url = "http://example.com",
                                 tp = "TPTPTP"
                               }
                  }
  let ctx = Pulumi c
  let o = fmap (\x ->  Oidc {
                   address = (url (identity x)),
                   thumbprint = (tp (identity x)),
                   Main.id = 2
                   }) ctx
  print $ show o
  {- result: -} 
  {- "Pulumi (Oidc {address = \"http://example.com\", thumbprint = \"TPTPTP\", id = 2})" -}

```

That is, within the Pulumi monad, pass data from the Cluster datatype to the Oidc datatype. (Apologies for the haskell, but the Pulumi design so strongly resembles a monad I felt it was most natural)

**Reply by rawkode** [reply]
> @rawkode - sure! Sorry about that.
> 
> ```
> oidc, err := aws.NewOpenIdConnectProvider(ctx, "Test1", &aws.OpenIdConnectProviderArgs{
> 			ClientIdLists: pulumi.StringArray{
> 				pulumi.String("sts.amazonaws.com"),
> 			},
> 			ThumbprintLists: tp,
> 			Url:             url,
> 		})
> ```

I believe passing the `AnyOutput` to that should work just fine. Did you get an error?

**Reply by pnathan** [reply]
Handling the anyoutput is extremely unclear - what's the stored type? what can you do with it? Casting generated errors. I've worked... around... that specific example, but the fundamental problem remains. 

Is AnyOutput notionally the top / bottom of the hierarchy?

**Reply by rawkode** [reply]
Sorry for the delay, was quite the week. I'll build a small repro of this tomorrow 👍
